pkg/storegateway/bucket_chunk_reader.go

```diff
@@ -126,88 +131,66 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesChunks,
 
-	var (
-		buf        = make([]byte, mimir_tsdb.EstimatedMaxChunkSize)
-		readOffset = int(pIdxs[0].offset)
-
-		// Save a few allocations.
-		written  int64
-		diff     uint32
-		chunkLen int
-		n        int
-	)
+	reader.Reset(uint64(pIdxs[0].offset), bucketReader)
+	defer reader.Release()
 
 	for i, pIdx := range pIdxs {
-		// Fast forward range reader to the next chunk start in case of sparse (for our purposes) byte range.
-		for readOffset < int(pIdx.offset) {
-			written, err = io.CopyN(io.Discard, bufReader, int64(pIdx.offset)-int64(readOffset))
-			if err != nil {
-				return errors.Wrap(err, "fast forward range reader")
-			}
-			readOffset += int(written)
-		}
-		// Use the chunk length estimation.
-		// However, declaration for length warns us this estimation can be wrong.
-		// This is handled further down below.
-		chunkLen = int(pIdx.length)
-		if i+1 < len(pIdxs) {
-			if diff = pIdxs[i+1].offset - pIdx.offset; int(diff) < chunkLen {
-				chunkLen = int(diff)
-			}
-		}
-		cb := buf[:chunkLen]
-		n, err = io.ReadFull(bufReader, cb)
-		readOffset += n
-		// Unexpected EOF for last chunk could be a valid case. Any other errors are definitely real.
-		if err != nil && !(errors.Is(err, io.ErrUnexpectedEOF) && i == len(pIdxs)-1) {
-			return errors.Wrapf(err, "read range for seq %d offset %x", seq, pIdx.offset)
+		// Fast-forward range reader to the next chunk start in case of sparse (for our purposes) byte range.
+		if err = reader.SkipTo(uint64(pIdx.offset)); err != nil {
+			return errors.Wrap(err, "fast forward range reader")
 		}
-
-		chunkDataLen, n := binary.Uvarint(cb)
-		if n < 1 {
-			return errors.New("reading chunk length failed")
-		}
-
-		// Chunk length is n (number of bytes used to encode chunk data), 1 for chunk encoding and chunkDataLen for actual chunk data.
-		// There is also crc32 after the chunk, but we ignore that.
-		chunkLen = n + 1 + int(chunkDataLen)
-		if chunkLen <= len(cb) {
-			err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), rawChunk(cb[n:chunkLen]), chunksPool)
-			if err != nil {
-				return errors.Wrap(err, "populate chunk")
-			}
-			localStats.chunksTouched++
-			localStats.chunksTouchedSizeSum += chunkLen + crc32.Size
-			continue
-		}
-
-		// Read entire chunk into new buffer.
-		// TODO: readChunkRange call could be avoided for any chunk but last in this particular part.
-		nb, err := r.block.readChunkRange(ctx, seq, int64(pIdx.offset), int64(chunkLen), []byteRange{{offset: 0, length: chunkLen}})
+		chunkDataLen, err := binary.ReadUvarint(reader)
 		if err != nil {
-			return errors.Wrapf(err, "preloaded chunk too small, expecting %d, and failed to fetch full chunk", chunkLen)
+			return errors.Wrap(err, "parsing reader")
 		}
-		if len(*nb) != chunkLen {
-			return errors.Errorf("preloaded chunk too small, expecting %d", chunkLen)
+		// We ignore the crc32 after the chunk data.
+		chunkEncDataLen := int(chunkDataLen) + chunks.ChunkEncodingSize
```

Comment:
[nit] Really a nit, but I would sort them by the same order of their fields in the serialised data.

```suggestion
		chunkEncDataLen := chunks.ChunkEncodingSize + int(chunkDataLen)
```